typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+| 6   | Get lowest address (end) of the grant region       | `()`     | `usize`? | `ErrorCode` |
+| 7   | Get num. writeable flash regions in process header | `()`     | `u32`    | `ErrorCode` |
+| 8   | Get start address of a writeable flash region      | `u32`    | `void *` | `ErrorCode` |
+| 9   | Get length of a writeable flash region             | `u32`    | `usize`  | `ErrorCode` |
+| 10  | Set the start of the process stack                 | `usize`  | `()`     | `ErrorCode` |
+| 11  | Set the start of the process heap                  | `usize`  | `()`     | `ErrorCode` |
+
+Note: there are several places where pointer versus usize can be bikeshed. Also,
+I did not fully adapt this for CHERI.
+
+### Exit
+
+Arguments:
+
+| Register | Argument        | Type  |
+| -------- | --------------- | ----- |
+| a1       | Exit number     | `u32` |
+| a2       | Completing code | `u32` |
+
+`exit-restart` and `exit-terminate` never return so return variants are not
+specified.
+
+## Userspace library implementation.
+
+I believe it is possible to use Rust's generics to provide nice APIs for
+invoking the above commands. For example, I believe `libtock-rs` could define
+a command function with the following signature:
+
+```rust
+// Implemented on any type that can be represented as an ArbitraryData(15)
+trait CommandReturn { ... }
+
+// Implemented on any type that can be represented as an ArbitraryData in a3-...
+trait CommandArgs { ... }
+
+fn command<Args: CommandArgs, Success: CommandReturn, Failure: CommandReturn>(
+    driver_num: u32,
+    command_num: u32,
+    args: Args) -> Result<Result<Success, Failure>, ErrorCode> { ... }
+```
+
+For libtock-c and other languages with less powerful generics, we would generate
+the interface code for each syscall driver.
+
+It is expected that userspace libraries will provide a way to register upcalls
+that have a function signature matching the upcall's concrete type instead of
+`ArbitraryData`; they'll wrap that function in one that decodes the
+`ArbitraryData`. The userspace library will pass the latter function to
+Subscribe.
+
+## Properties
+
+This pros and cons list is specified relative to Tock 2.0's syscall ABI.
+
+Pros:
+
+1. No more manually casting everything to a `u32`/`usize` to invoke Command or
+   an upcall; any casting necessary will be handled by the core syscall layer
+   (or code generator).
+1. Type safety: If there is a type mismatch between what userspace expects a
+   syscall to use and what the syscall actually uses, that will be caught
+   (by the kernel for arguments and by the userspace library for return values)
+   and result in an error code rather than data corruption.
+1. Argument count checks: no more `command(..., ..., ..., 0 /* unused */)`;
+   syscall invocations specify exactly the number of arguments the system call
+   needs. The same applies to upcalls.
+1. Command can accept many more arguments and return many more values in both
+   its success and error case.
```

Comment:
To me, the obvious huge pro here is the flexibility and expressibility gains for passing data, which TRD104 is much more limited on. Any arbitrary collection of useful data types could be specified, with clear semantics.

Reply:
Yes. I added your last sentence to this bullet point.